### PR TITLE
Add option for SSL cert algorithm in DuckDNS addon

### LINKED ISSUE #2183: DuckDNS add-on generates certificates that are incompatible with Android 7.0
## The problem

The Let's encrypt SSL certificates that are generated with the current version of the DuckDNS add-on (1.13.0) are not compatible with Android 7.0 which breaks the submission of sensor data to home-assistant from the companion app: https://github.com/home-assistant/android/issues/1709

The dehydrated script has an option to select the algorithm that is used for generating the certificate: 
```
--algo (-a) rsa|prime256v1|secp384r1 Which public key algorithm should be used? Supported: rsa, prime256v1 and secp384r1
```

I changed the the duckDNS add-on locally to use the `prime256v1` algorithm instead of the `secp384r1` (default) which solved the issue on my setup. 

I suggest to expose this command line option through the add-on configuration to facilitate adjustments of the certificate algorithm.

## Environment

- Add-on with the issue: DuckDNS
- Add-on release with the issue: 1.13.0
- Last working add-on release (if known): unknown
- Operating environment (OS/Supervised): Home Assistant Supervised on Debian

## Problem-relevant configuration

```yaml
lets_encrypt:
  accept_terms: true
  certfile: fullchain.pem
  keyfile: privkey.pem
token: '!secret duckdnstoken'
domains:
  - '!secret duckdnsurl'
aliases: []
seconds: 300
```

## Traceback/Error logs

Relevant logs from the companion app are included in this issue: https://github.com/home-assistant/android/issues/1709

## Additional information

I can try to open a pull-request with relevant changes but i anticipate that the exact implementation might need some discussion, thus opening this issue first.

### PR DESCRIPTION
**Pull request by tux2000**
* Add optional parameter `lets_encrypt.algo` in `config.json`
* Use parameter in `run.sh` as commandline option to dehydrated
* Update documentation
* Bump version

Closes #2183

### CODE REVIEW

**Comment by sstefanowski** [comment]
I vote for this pull request :)
It will also fix problems with Alexa playback of MP3s stored in HA (exposes via HTTPS and played using SSML Audio tag).
Currently, Alexa cannot access MP3 as it cannot make a handshake with HA secured with default algorithm of  certificate generation.

**Comment by sstefanowski** [comment]
I vote for this pull request :)
It will also fix problems with Alexa playback of MP3s stored in HA (exposed via HTTPS and played using SSML Audio tag).
Currently, Alexa cannot access MP3 as it cannot make a handshake with HA secured by a certificate generated with the default algorithm.

@ludeeus 
Hi, Could you please consider the approval of this pull request? 
I checked the commit history... and it looks like in July you have committed the change including an upgrade of the dehydrated version from 0.6.5 to 0.7.0. 
Dehydrated 0.6.5 used `rsa` as the default "certificate algorithm". Dehydrated 0.7.0 uses `secp384r1`. 
Since that time some clients (also important like Amazon Alexa SSML engine) cannot do the HTTPS handshake. It can be successful if a certificate could be generated using another algorithm (e.g. `rsa`)

File: duckdns/data/run.sh [hunk 1]
```diff
@@ -12,6 +12,7 @@ if bashio::config.has_value "ipv6"; then IPV6=$(bashio::config 'ipv6'); else IPV
 TOKEN=$(bashio::config 'token')
 DOMAINS=$(bashio::config 'domains | join(",")')
 WAIT_TIME=$(bashio::config 'seconds')
+if bashio::config.has_value "algo"; then ALGO=$(bashio::config 'algo'); else ALGO="secp384r1"; fi
```
Comment (ludeeus) [inline duckdns/data/run.sh:15]:
```suggestion
if bashio::config.has_value "lets_encrypt.algo"; then ALGO=$(bashio::config 'lets_encrypt.algo'); else ALGO="secp384r1"; fi
```

File: duckdns/config.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "Duck DNS",
-  "version": "1.13.0",
+  "version": "1.14.0",
```
Comment (ludeeus) [inline duckdns/config.json:3]: Changelog is missing an update

File: duckdns/DOCS.md [hunk 3]
```diff
@@ -62,6 +62,11 @@ The name of the private key file generated by Let's Encrypt. The private key fil
 
 **Note**: _The file is stored in `/ssl/`, which is the default for Home Assistant_
 
+#### Option `lets_encrypt.algo` (optional)
+
+Which public key algorithm should be used? Supported: `rsa`, `prime256v1` and `secp384r1`. Default is `secp384r1`, other algorithms might improve compatibility with older devices.
```
Comment (ludeeus) [inline duckdns/DOCS.md:67]:
```suggestion
Public key algorithm that will be used.

Supported values: `rsa`, `prime256v1` and `secp384r1`. 

The default is `secp384r1`
```

File: duckdns/config.json [hunk 4]
```diff
@@ -22,7 +23,8 @@
     "lets_encrypt": {
       "accept_terms": "bool",
       "certfile": "str",
-      "keyfile": "str"
+      "keyfile": "str",
+      "algo": "str?"
```
Comment (ludeeus) [inline duckdns/config.json:27]:
```suggestion
      "algo": "list(rsa|prime256v1|secp384r1)?"
```

**Review by ludeeus — CHANGES_REQUESTED** [review]

File: duckdns/config.json (hunk 2 above)
Comment (tux2000) [reply duckdns/config.json:3]: I will add it

File: duckdns/config.json [hunk 5]
```diff
@@ -22,7 +23,8 @@
     "lets_encrypt": {
       "accept_terms": "bool",
       "certfile": "str",
-      "keyfile": "str"
+      "keyfile": "str",
+      "algo": "list(rsa|prime256v1|secp384r1)?"
```
Comment (pvizeli) [inline duckdns/config.json:27]:
```suggestion
      "algo": "list(rsa|prime256v1|secp384r1)"
```

File: duckdns/data/run.sh [hunk 6]
```diff
@@ -12,6 +12,7 @@ if bashio::config.has_value "ipv6"; then IPV6=$(bashio::config 'ipv6'); else IPV
 TOKEN=$(bashio::config 'token')
 DOMAINS=$(bashio::config 'domains | join(",")')
 WAIT_TIME=$(bashio::config 'seconds')
+if bashio::config.has_value "lets_encrypt.algo"; then ALGO=$(bashio::config 'lets_encrypt.algo'); else ALGO="secp384r1"; fi
```
Comment (pvizeli) [inline duckdns/data/run.sh:15]:
```suggestion
ALGO=$(bashio::config 'lets_encrypt.algo')
```

**Review by pvizeli — APPROVED** [review]

**Review by pvizeli — APPROVED** [review]

**Comment by teknomar7** [comment]
The "algo" parameter does not appear to be optional as indicated in the documentation.  When I updated this morning, duckdns would not start back up without defining an algo k/v.  

This was seen in the logs prior to defining the k/v in the configuration (token and dns names have been redacted):
```
21-09-27 08:30:30 ERROR (MainThread) [supervisor.addons.addon] Add-on core_duckdns has invalid options: Missing option 'algo' in lets_encrypt in Duck DNS (core_duckdns). Got {'lets_encrypt': {'accept_terms': True, 'certfile': 'fullchain.pem', 'keyfile': 'privkey.pem'}, 'token': '<redacted>', 'domains': ['<redacted>'], 'aliases': [], 'seconds': 300}
```

**Comment by btehan** [comment]
Same as @teknomar7.

**Comment by tux2000** [comment]
The changes in https://github.com/home-assistant/addons/pull/2189/commits/dac5764aebe38f891456be7300f28678877eb7bd made the option non-optional. 
Not sure what works best, updating the documentation or making the option optional again. I guess forcing all users out there to manually update the configuration will not be optimal...

**Comment by teknomar7** [comment]
I would think you'd want to have the default value set if not specified.  It lists the default value in the documentation as secp384r1, so it seems like not having a default value was the mistake here (not the incorrect documentation).  Furthermore, this wasn't a major version bump, which would imply it's not a breaking change per semantic versioning when this does break functionality without some type of human interaction.

**Comment by tux2000** [comment]
Yes, this was exactly my reasoning when implementing this parameter as optional. Not sure if something with the implementation as it was in https://github.com/home-assistant/addons/pull/2189/commits/b2f282c33bab8288ee91e843e3cb7f42860c7d2c did not work or what the reason was that @pvizeli changed the behavior but not the version bump or the documentation prior to merging this to main.

**Comment by esenterre** [comment]
Well. I had the problem with the upgrade that "algo" is not optional. I don't care adding line to teh config. But I don't know what (sorry for being dummy :) )

For now, I added :
`algo: secp384r1`
But should it be 
`algo: rsa`
 ?

**Comment by teknomar7** [comment]
> For now, I added :
> `algo: secp384r1`
> But should it be
> `algo: rsa`
> ?

There isn't a right or wrong answer here.  I'd go with the default secp384r1, but if you have some client devices that call your https URL that stop working, see if there are any SSL handshake issues in the logs... which might mean you'd need to switch to RSA.  The default secp384r1 is ECC and is supposed to improve performance over RSA from my understanding, but might not work on older devices that never got upgraded to work with that encryption type.  Most people probably won't notice the difference between any of the options to be honest.

**Pull request merged**